src/data-final-fixes.lua

```diff
@@ -1,49 +1,105 @@
--- Hello there :)
-local fluidTech = data.raw.technology['fluid-handling']
+--[[
 
--- https://forums.factorio.com/viewtopic.php?f=23&t=100057
-local targetTechs = {
-  -- base game
-  ['sulfuric-acid-barrel'] = 'sulfur-processing',
-  ['empty-sulfuric-acid-barrel'] = 'sulfur-processing',
+Iterate over all existing fluids and their barrel/empty barrel corresponding techs and
+ensure that their icons are not shown until the related fluid tech has been researched.
 
-  ['crude-oil-barrel'] = 'oil-gathering',
-  ['empty-crude-oil-barrel'] = 'oil-gathering',
+NOTE: If a barrel does not show up, it's because it's not properly named according to what
+Factorio expects. Basically, the gist is that the fluid should be named like the
+following:
 
-  ['petroleum-gas-barrel'] = 'oil-processing',
-  ['empty-petroleum-gas-barrel'] = 'oil-processing',
+- <fluidName>-barrel
+- empty-<fluidName>-barrel
 
-  ['heavy-oil-barrel'] = 'advanced-oil-processing',
-  ['empty-heavy-oil-barrel'] = 'advanced-oil-processing',
+--]]
 
-  ['light-oil-barrel'] = 'advanced-oil-processing',
-  ['empty-light-oil-barrel'] = 'advanced-oil-processing',
+local technologyData = data.raw["technology"]
+local recipeData = data.raw["recipe"]
+local fluidTech = technologyData["fluid-handling"]
 
-  ['lubricant-barrel'] = 'lubricant',
-  ['empty-lubricant-barrel'] = 'lubricant',
+---@type table<string, string>
+local fluidToTechMapping = {}
 
-  -- space age
-  ['fluoroketone-cold-barrel'] = 'cryogenic-plant',
-  ['empty-fluoroketone-cold-barrel'] = 'cryogenic-plant',
+-- Logs to the mod file for convenience purposes. Contents are added to "script-output/barrel-stages.log"
+---@param msg string
+local function modLog(msg)
+  helpers.write_file("barrel-stages.log", msg, true)
+end
 
-  ['fluoroketone-hot-barrel'] = 'cryogenic-plant',
-  ['empty-fluoroketone-hot-barrel'] = 'cryogenic-plant',
-}
+-- Helper function to convert a table to a string
+---@param tableToDump table - The table that you want to conver to a string
+---@return string - The table, dumped as an ugly string
+local function dumpTable(tableToDump)
+  if type(tableToDump) == "table" then
+    local s = "{ "
+    for k, v in pairs(tableToDump) do
+      if type(k) ~= "number" then
+        k = '"' .. k .. '"'
+      end
+      s = s .. "[" .. k .. "] = " .. dumpTable(v) .. ","
+    end
+    return s .. "} "
+  else
+    return tostring(tableToDump)
+  end
+end
```

Comment:
I reckon something like that exists: https://lua-api.factorio.com/latest/classes/LuaHelpers.html#table_to_json